File: Javascript/Hashmap/M-1657. Determine if Two Strings Are Close.ts
function closeStrings(word1: string, word2: string): boolean {
  if (word1.length !== word2.length) return false;
  const word1Map = new Map<string, number>();
  const word2Map = new Map<string, number>();
  for (const char of word1) word1Map.set(char, (word1Map.get(char) || 0) + 1);
  for (const char of word2) word2Map.set(char, (word2Map.get(char) || 0) + 1);

  const word1DistChars = Array.from(word1Map.keys()).sort().join('');
  const word2DistChars = Array.from(word2Map.keys()).sort().join('');
  // console.log('word1DistChars', word1DistChars);
  // console.log('word2DistChars', word2DistChars);
  if (word1DistChars !== word2DistChars) return false;

  const word1CharCount = Array.from(word1Map.values()).sort((a, b) => a - b);
  const word2CharCount = Array.from(word2Map.values()).sort((a, b) => a - b);
  // console.log('word1CharCount', word1CharCount);
  // console.log('word2CharCount', word2CharCount);
  const isCharCountEqual = word1CharCount.every(
    (value, index) => value === word2CharCount[index]
  );
  // console.log('isCharCountEqual', isCharCountEqual);
  return isCharCountEqual;
}

const testData = [
  {
    word1: 'abc',
    word2: 'bca',
    ans: true,
  },
  {
    word1: 'a',
    word2: 'aa',
    ans: false,
  },
  {
    word1: 'cabbba',
    word2: 'abbccc',
    ans: true,
  },
];

for (const data of testData) {
  console.log(data['ans'] === closeStrings(data['word1'], data['word2']));
}

File: Python/Hashmap/M-1657. Determine if Two Strings Are Close.py
class Solution:
    @classmethod
    def closeStrings(self, word1: str, word2: str) -> bool:
        if len(word1) != len(word2):
            return False

        word1_map = {}
        word2_map = {}
        for char in word1:
            word1_map[char] = (word1_map.get(char) or 0) + 1
        for char in word2:
            word2_map[char] = (word2_map.get(char) or 0) + 1

        # compare chars
        word1_dist_chars = list(word1_map.keys())
        word2_dist_chars = list(word2_map.keys())
        word1_dist_chars.sort()
        word2_dist_chars.sort()

        if word1_dist_chars != word2_dist_chars:
            return False

        # compare occurrence
        word1_char_cnt = list(word1_map.values())
        word2_char_cnt = list(word2_map.values())
        word1_char_cnt.sort()
        word2_char_cnt.sort()
        return word1_char_cnt == word2_char_cnt


testData = [
    {
        "word1": "abc",
        "word2": "bca",
        "ans": True,
    },
    {
        "word1": "a",
        "word2": "aa",
        "ans": False,
    },
    {
        "word1": "cabbba",
        "word2": "abbccc",
        "ans": True,
    },
]

for data in testData:
    print(data["ans"] == Solution.closeStrings(data["word1"], data["word2"]))
